### PR TITLE
okhttp: add full implementation of HPACK header compression

### LINKED ISSUE #5985: OkHttp HPACK header compression is not fully implemented
Disregarding Huffman encoding, upstream Okhttp's HPACK implementation does not fully implement HPACK compression. Especially for the [writer](https://github.com/grpc/grpc-java/blob/8a9afd618aec0e1bf5858232cac70c89a90fb96e/okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java#L383), its implementation is very minimal, with very simplistic [header field representations](https://tools.ietf.org/html/rfc7541#section-6) (for table lookup, `Indexed Header Field Representation` is completely not implemented, only for indexed name; no dynamic table is involved at all).

An obvious result is that in Http2 frames, header fields such as `:scheme: https` will be encoded as `06 05 68 74 70 73` and `:method: POST` will be encoded as `02 04 50 4f 53 54`, which are in the format for `Literal Header Field without Indexing - Indexed Name` although `:scheme: https` and `:method: POST` do exist in the static table.

[Okhttp3](https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/internal/http2/Hpack.kt) has fully implemented this.

### PR DESCRIPTION
**Pull request by voidzcy**
Copied implementation from [OkHttp3 Hpack.kt](https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/internal/http2/Hpack.kt). Most changes are on Writer.

Tests are copied from [OkHttp3 HpackTest.java](https://github.com/square/okhttp/blob/master/okhttp/src/test/java/okhttp3/internal/http2/HpackTest.java), but I was trying to avoid introducing dependency of `google/Truth` by preserving the usage of `org.junit.Assert.assertEquals`.

Upstream okhttp3 has pretty good coverage for unit tests. I also inspected packets after the change:

- Before:

okhttp_before

`:scheme: https`, `:method: POST` and `:authority: localhost:8080` all represented in the format of `Literal Header Field without indexing - Indexed Name`.

- After:

okhttp_after

`:scheme: https` and `:method: POST` are represented in the format of `Indexed Header Field` and `:authority: localhost:8080` in the format of `Literal Header Field with Incremental Indexing - Indexed Name`.

- As a reference, packets from Netty's:

netty

Resolves #5985.

### CODE REVIEW

File: okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Header.java [hunk 1]
```diff
@@ -9,6 +9,8 @@
 /** HTTP header: the name is an ASCII string, but the value can be UTF-8. */
 public final class Header {
   // Special header names defined in the SPDY and HTTP/2 specs.
+  static final ByteString PSEUDO_PREFIX = ByteString.encodeUtf8(":");
```
Comment (dapengzhang0) [inline okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Header.java]: This should be moved to Hpack.java

File: okhttp/third_party/okhttp/test/java/io/grpc/okhttp/internal/framed/HpackTest.java [hunk 2]
```diff
@@ -132,33 +140,84 @@
...
+    // Simulate receiving a small dynamic table size update, that implies eviction.
+    bytesIn.writeByte(0x3F); // Dynamic table size update (size = 55).
+    bytesIn.writeByte(0x18);
+    hpackReader.readHeaders();
     assertEquals(1, hpackReader.dynamicTableHeaderCount);
   }
 
   /** Header table backing array is initially 8 long, let's ensure it grows. */
   @Test public void dynamicallyGrowsBeyond64Entries() throws IOException {
+    // Lots of headers need more room!
```
Comment (dapengzhang0) [inline okhttp/third_party/okhttp/test/java/io/grpc/okhttp/internal/framed/HpackTest.java]: I would not change the existing test and only add new tests for new features. Will they break if not changed?

File: okhttp/third_party/okhttp/test/java/io/grpc/okhttp/internal/framed/HpackTest.java [hunk 3]
```diff
@@ -748,7 +1147,7 @@ private ByteString intArrayToByteArray(int[] bytes) {
     return ByteString.of(data);
   }
 
-  private int headerTableLength() {
+  private int readerHeaderTableLength() {
```
Comment (dapengzhang0) [inline okhttp/third_party/okhttp/test/java/io/grpc/okhttp/internal/framed/HpackTest.java]: We should not refactor name.

File: okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java [hunk 4]
```diff
@@ -373,26 +389,105 @@ ByteString readByteString() throws IOException {
...
+     */
+    private int smallestHeaderTableSizeSetting = Integer.MAX_VALUE;
+    private boolean emitDynamicTableSizeUpdate = false;
+    private int maxDynamicTableByteCount;
+
+    // Visible for testing.
+    io.grpc.okhttp.internal.framed.Header[] dynamicTable = new io.grpc.okhttp.internal.framed.Header[8];
+    // Array is populated back to front, so new entries always have lowest index.
+    int nextDynamicTableIndex = dynamicTable.length - 1;
+    int dynamicTableHeaderCount = 0;
```
Comment (dapengzhang0) [inline okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java]: nit: this is `headerCount` in Hpack.kt

Comment (dapengzhang0) [inline okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java]: nit: no need to initialize `0` or `false` values.

File: okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java [hunk 5]
```diff
@@ -373,26 +389,105 @@ ByteString readByteString() throws IOException {
...
+     * header blocks, we need to keep track of the smallest value in that interval.
+     */
+    private int smallestHeaderTableSizeSetting = Integer.MAX_VALUE;
+    private boolean emitDynamicTableSizeUpdate = false;
+    private int maxDynamicTableByteCount;
+
+    // Visible for testing.
+    io.grpc.okhttp.internal.framed.Header[] dynamicTable = new io.grpc.okhttp.internal.framed.Header[8];
+    // Array is populated back to front, so new entries always have lowest index.
+    int nextDynamicTableIndex = dynamicTable.length - 1;
```
Comment (dapengzhang0) [inline okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java]: nit: this `nextHeaderIndex` in Hpack.kt

File: okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java (hunk 4 above)
Comment (dapengzhang0) [inline okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java]: nit: this field is private and has a package-private/VisibleForTesting getter, that is, can not be mutated outside the class.

File: okhttp/third_party/okhttp/test/java/io/grpc/okhttp/internal/framed/HpackTest.java [hunk 6]
```diff
@@ -49,7 +49,7 @@
 
   @Before public void reset() {
     hpackReader = newReader(bytesIn);
-    hpackWriter = new Hpack.Writer(bytesOut);
+    hpackWriter = new Hpack.Writer(4096, false, bytesOut);
```
Comment (dapengzhang0) [inline okhttp/third_party/okhttp/test/java/io/grpc/okhttp/internal/framed/HpackTest.java]: I would do this only for new tests that need it.

Comment (voidzcy) [reply okhttp/third_party/okhttp/test/java/io/grpc/okhttp/internal/framed/HpackTest.java]: Previously the default is to not use compression and no dynamic table (i.e., not table size setting), but now we have these definitions, even for existing tests that do not rely on them.

File: okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Header.java (hunk 1 above)
Comment (voidzcy) [reply okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Header.java]: Done.

File: okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java (hunk 4 above)
Comment (voidzcy) [reply okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java]: Done.

Comment (voidzcy) [reply okhttp/third_party/okhttp/main/java/io/grpc/okhttp/internal/framed/Hpack.java]: Done.

**Comment by vkryachko** [comment]
Hi, when is this expected to be merged? And subsequently released?

**Comment by voidzcy** [comment]
> This looks fine to me, although I would like to see what the second RPC looks like with huffman disabled. In general it may be reasonable to disable huffman.


Here is the result (the main gRPC frame) of running the interop-testing client with two successive empty call. Huffman encoding was disable (but the default is enabled in implementation).

- First:
okhttp_test1

- Second:
okhttp_test2

**Comment by voidzcy** [comment]
> I still think it would be good to disable huffman. But this looks fine.

Alright, I will disable huffman. Btw, may I ask why?

**Comment by ejona86** [comment]
It's a CPU vs bandwidth argument. The most compact huffman code is 5 bits long. So the theoretical maximum bandwidth reduction it could make is 37%, but it will generally be less than that. Also, for short strings (say 10 bytes) who cares if you save a few bytes? Your overhead is much higher than that. So we could enable it on just larger strings, but many of the long strings we tend to send are also those we only send once, like user-agent. Although grpc-trace could benefit a bit. I'd be fine enabling it for strings 32+ bytes or some such.

**Pull request merged**